lib/dynamoid/transaction/update.rb

```diff
@@ -0,0 +1,68 @@
+require_relative 'action'
+
+module Dynamoid
+  module Transaction
+    class Update < Action
+
+      def run_callbacks
+        (yield; return) unless model
+        model.run_callbacks(:save) do
+          model.run_callbacks(:update) do
+            model.run_callbacks(:validate) do
+              yield
+            end
+          end
+        end
+      end
+
+      def to_h
+        item = attributes.clone
+
+        # ensure types are correct for incoming attributes
+        all_types_string = item.keys.map{|k| [k,{type: :string}]}.to_h
+        item = add_timestamps(item, skip_created_at: true)
+        item = model.attributes.merge(item) if model
+        # TODO only update dirty attributes, deleting the inheritance field will not be needed
+        item.delete(model_class.inheritance_field) # do not update type field
+        item = Dynamoid::Dumping.dump_attributes(item, all_types_string.merge(model_class.attributes))
+
+        # set 'key' that is used to look up record for updating
+        hash_key = item.delete(model_class.hash_key)
+        key = {model_class.hash_key => hash_key}
+        key[model_class.range_key] = item.delete(model_class.range_key) if model_class.range_key?
+
+        # e.g. "SET #updated_at = :updated_at ADD record_count :i"
+        update_expression = options[:update_expression] || ("SET " + item.keys.map{ |k| "##{k} = :#{k}" }.join(", "))
+
+        # e.g. {":updated_at" => 1645453.234, ":i" => 1}
+        expression_attribute_values = item.map{|k,v| [":#{k}",v]}.to_h
+
+        if options[:add]
+          update_expression += " ADD " + options[:add].keys.map{|k| "#{k} :#{k}"}.join(", ")
+          options[:add].each{|k,v| expression_attribute_values[":#{k}"] = v }
+        end
+
+        # only alias names for fields in models, other values such as for ADD do not have them
+        # e.g. {"#updated_at" => "updated_at"}
+        attribute_keys_in_model = item.keys.intersection(model_class.attributes.keys)
+        expression_attribute_names = options[:expression_attribute_names] || attribute_keys_in_model.map{|k| ["##{k}","#{k}"]}.to_h
+
+        condition_expression = "attribute_exists(#{model_class.hash_key})" # fail if record is missing
+        condition_expression += " and attribute_exists(#{model_class.range_key})" if model_class.range_key? # needed?
+
+        result = {
+          update: {
+            key: key,
+            table_name: model_class.table_name,
+            update_expression: update_expression,
+            expression_attribute_values: expression_attribute_values
+          }
+        }
+        result[:update][:expression_attribute_names] = expression_attribute_names if expression_attribute_names.present?
+        result[:update][:condition_expression] = condition_expression unless item[:skip_existence_check]
+        puts "UPDATE #{result}"
```

Comment:
For debugging all requests/responses to DynamoDb could be logged with a special formatter and setting log level to `debug` like it's done in tests

https://github.com/Dynamoid/dynamoid/blob/6d53f100f0478bc04c7c45646771182df603fdc8/spec/spec_helper.rb#L76

https://github.com/Dynamoid/dynamoid/blob/6d53f100f0478bc04c7c45646771182df603fdc8/spec/support/log_level.rb#L3-L11

Reply:
Thanks. I removed the puts.